src/libopenrave/libopenrave.h

```diff
@@ -481,6 +481,152 @@ void UpdateOrCreateInfo(const rapidjson::Value& value, std::vector<boost::shared
+            }
+
+            // If we are updating an element in place, we need to check if we're about to change the id/name
+            // If we are, we have to update the jump tables.
+            const bool didChangeName = !name.empty() && vInfosOut[existingInfoIndex]->_name != name; // If the name to be loaded is empty, it won't clobber anything, so don't count that as a change
+            const bool didChangeId = vInfosOut[existingInfoIndex]->_id != id; // Id always overwrites, so count it as changed even if the new id is empty
+
+            // Erase the old id/name mapping if it changed
+            if (didChangeId) {
+                infoIdxById.erase(vInfosOut[existingInfoIndex]->_id);
+            }
+            if (didChangeName) {
+                infoIdxByName.erase(vInfosOut[existingInfoIndex]->_name);
+            }
+
+            // Deserialize the new json over the existing entry
+            vInfosOut[existingInfoIndex]->DeserializeJSON(value, fUnitScale, options);
+            vInfosOut[existingInfoIndex]->_id = id;
+
+            // Write back the new id/name mappings with the new values if changed
+            if (didChangeId) {
+                infoIdxById.emplace(vInfosOut[existingInfoIndex]->_id, existingInfoIndex);
+            }
+            if (didChangeName) {
+                infoIdxByName.emplace(vInfosOut[existingInfoIndex]->_name, existingInfoIndex);
+            }
+
+            continue;
+        }
+
+        // If the item is deleted and didn't match an existing index, then there is nothing to remove.
+        // Just skip this entry.
+        if (isDeleted) {
+            continue;
+        }
+
+        // If we didn't match an existing entry, then create a new one now
+        boost::shared_ptr<T> pNewInfo(new T());
+        pNewInfo->DeserializeJSON(value, fUnitScale, options);
+        pNewInfo->_id = id;
+
+        // Update our jump tables and push this info onto the list
+        const size_t newInfoIndex = vInfosOut.size();
+        infoIdxById.emplace(id, newInfoIndex);
+        infoIdxByName.emplace(pNewInfo->_name, newInfoIndex);
+        vInfosOut.emplace_back(std::move(pNewInfo));
+    }
+
+    // If we deleted any elements, scan through and ensure there are no holes in the output before returning.
+    // Scan backwards through the list, swapping all of the invalid pointers to the end.
+    // Once we're done, truncate the list to drop all the invalid entries.
+    if (didEraseElements) {
+        size_t validElementCount = vInfosOut.size();
+        for (ssize_t i = vInfosOut.size() - 1; i >= 0; i--) {
+            // If this element is still valid, leave it where it is
+            if (!!vInfosOut[i]) {
+                continue;
+            }
+
+            // If this element is null, swap it with the last valid element and decrement the valid element count
+            std::swap(vInfosOut[i], vInfosOut[--validElementCount]); // Note prefix decrement to turn the count into an index
```

Comment:
just curious here, why can't we use a simple assignment and have to `swap()` ?

Reply:
Swapping is faster because it avoids a construct/incref (when assigning the last valid element into this position) and a desctruct/decref (when we truncate the array). Shared pointer refcounting is atomic, so it's a bit expensive, and swapping/moving where possible helps avoid this cost.

Reply:
sounds good, thanks!